lib/Serialization/SerializeSIL.cpp

```diff
@@ -1888,6 +1888,25 @@ void SILSerializer::writeSILInstruction(const SILInstruction &SI) {
         UI->hasOperand() ? addValueRef(UI->getOperand()) : ValueID());
     break;
   }
+  case SILInstructionKind::CXXVirtualMethodInst: {
+    // Format: a type, an operand and a SILDeclRef. Use SILOneTypeValuesLayout:
+    // type, Attr, SILDeclRef (DeclID, Kind, uncurryLevel),
+    // and an operand.
+    const CXXVirtualMethodInst *VMI = cast<CXXVirtualMethodInst>(&SI);
+    SILType Ty = VMI->getExtractedMethod()->getType();
+    SmallVector<ValueID, 9> ListOfValues;
+    auto operand = VMI->getOperand();
+    handleSILDeclRef(S, VMI->getMember(), ListOfValues);
+    ListOfValues.push_back(S.addTypeRef(operand->getType().getASTType()));
+    ListOfValues.push_back((unsigned)operand->getType().getCategory());
+    ListOfValues.push_back(addValueRef(operand));
+
+    SILOneTypeValuesLayout::emitRecord(
+        Out, ScratchRecord, SILAbbrCodes[SILOneTypeValuesLayout::Code],
+        (unsigned)SI.getKind(), S.addTypeRef(Ty.getASTType()),
+        (unsigned)Ty.getCategory(), ListOfValues);
```

Comment:
I don't think this serialization matches up with the deserialization on the other side. What are you actually trying to serialize? Why isn't it SILOneValueOneOperandLayout?

Reply:
The new instruction is identical in every way to ObjCMethodInst except that it also returns an adjusted this ptr as the second argument. I basically duplicated that logic for both the Deserialize and the sil printing. If you note, there are multiple instructions that follow this pattern and for parsing they all are packed into the same switch case which builds the right instruction after the parsing is done.

Reply:
Hm, okay. I feel like there should be an easier way but maybe that can wait for a general SIL serialization rework.